src/Symfony/Component/DependencyInjection/Compiler/ResolveDefinitionTemplatesPass.php

```diff
@@ -35,44 +36,78 @@ class ResolveDefinitionTemplatesPass implements CompilerPassInterface
      */
     public function process(ContainerBuilder $container)
     {
-        $this->container = $container;
         $this->compiler = $container->getCompiler();
         $this->formatter = $this->compiler->getLoggingFormatter();
 
         foreach ($container->getDefinitions() as $id => $definition) {
+            $this->currentId = $id;
+
             // yes, we are specifically fetching the definition from the
             // container to ensure we are not operating on stale data
             $definition = $container->getDefinition($id);
-            if (!$definition instanceof DefinitionDecorator || $definition->isAbstract()) {
-                continue;
+            if ($definition instanceof DefinitionDecorator) {
+                $definition = $this->resolveDefinition($container, $definition);
+                $container->setDefinition($id, $definition);
             }
 
-            $this->resolveDefinition($id, $definition);
+            $definition->setArguments($this->resolveArguments($container, $definition->getArguments()));
+            $definition->setMethodCalls($this->resolveArguments($container, $definition->getMethodCalls()));
+            $definition->setProperties($this->resolveArguments($container, $definition->getProperties()));
         }
     }
 
     /**
+     * Resolves definition decorator arguments.
+     *
+     * @param ContainerBuilder $container The ContainerBuilder
+     * @param array            $arguments An array of arguments
+     *
+     * @return array
+     */
+    private function resolveArguments(ContainerBuilder $container, array $arguments)
+    {
+        foreach ($arguments as $k => $argument) {
+            if (is_array($argument)) {
+                $arguments[$k] = $this->resolveArguments($container, $argument);
+            } elseif ($argument instanceof Definition) {
+                if ($argument instanceof DefinitionDecorator) {
+                    $arguments[$k] = $argument = $this->resolveDefinition($container, $argument);
+                }
+                $argument->setArguments($this->resolveArguments($container, $argument->getArguments()));
+                $argument->setMethodCalls($this->resolveArguments($container, $argument->getMethodCalls()));
+                $argument->setProperties($this->resolveArguments($container, $argument->getProperties()));
```

Comment:
The InlineServiceDefinitionsPass does not try to avoid duplicating the loop. I tried before pushing this one, by they are not the same (e.g. https://github.com/symfony/symfony/pull/15096/files#diff-d81a6a2f4968eb3c7d1511f42bdb8c79R43) even if they look close. I'd prefer keeping them separated, the code is more readable.